File: sp/reversi/server/src/main/cpp/reversi/engine/game_engine.cpp
#include "game_engine.hpp"

void game_engine::initialize_board(b_size init_x, b_size init_y) {
    if (game_board->get_cols() <= (init_x + 1) || game_board->get_rows() <= (init_y + 1)) {
        //TODO: exception throwing on invalid init
        return;
    }
    game_board->set_at(init_x, init_y, WHITE_PLAYER);
    game_board->set_at(init_x + 1, init_y, BLACK_PLAYER);
    game_board->set_at(init_x, init_y + 1, BLACK_PLAYER);
    game_board->set_at(init_x + 1, init_y + 1, WHITE_PLAYER);
}

size_t game_engine::get_possible_moves_count(player_code player) {
    size_t moves_count = 0;
    auto available_moves = _get_possible_moves(moves_count, player);
    return moves_count;
}

std::vector<bool> game_engine::_get_possible_moves(size_t &moves_count, player_code player) {
    size_t x;
    size_t y;
    auto moves = std::vector<bool>(game_board->get_cols() * game_board->get_cols());
    moves_count = 0;
    for (x = 0; x < this->game_board->get_cols(); ++x) {
        for (y = 0; y < this->game_board->get_rows(); ++y) {
            if (_is_valid_move(x, y, player)) {
                moves[y * this->game_board->get_cols() + x] = true;
                ++moves_count;
            }
        }
    }
    return moves;
}

player_code game_engine::_get_opponent_code(player_code player) {
    return (player == BLACK_PLAYER) ? WHITE_PLAYER : BLACK_PLAYER;
}

size_t game_engine::_is_valid_direction(b_size x, b_size y, int dir_x, int dir_y, player_code player) {
    player_code opponent = _get_opponent_code(player);
    size_t steps = 1;
    b_size pos_x;
    b_size pos_y;
    player_code cell;
    /*Check if move is in bounds and an opponent cell*/
    if ((x + dir_x >= this->game_board->get_cols()) ||
        (y + dir_y >= this->game_board->get_rows() || this->game_board->get_at(x + dir_x, y + dir_y) != opponent)) {
        return 0;
    }
    pos_x = x + steps * dir_x;
    pos_y = y + steps * dir_y;
    /*Check specified direction for a valid sequence of opponent cells*/
    while ((pos_x < this->game_board->get_cols()) && (pos_y < this->game_board->get_rows())) {
        cell = this->game_board->get_at(pos_x, pos_y);
        /*Empty cell -> move is invalid*/
        if (cell == NO_PLAYER) {
            return 0;
        }
        /*Player cell -> return number of steps (cells) in between*/
        if (cell == player) {
            return steps;
        }
        steps++;
        /*Calculate next cell for check*/
        pos_x = x + steps * dir_x;
        pos_y = y + steps * dir_y;
    }
    return 0;
}

bool game_engine::_is_valid_move(b_size x, b_size y, player_code player) {
    int dir_x;
    int dir_y;
    if (x >= this->game_board->get_cols() || y >= this->game_board->get_rows()) {
        return false;
    }
    if (this->game_board->get_at(x, y) != NO_PLAYER) {
        return false;
    }
    /*Check throw all possible directions*/
    for (dir_x = -1; dir_x <= 1; ++dir_x) {
        for (dir_y = -1; dir_y <= 1; ++dir_y) {
            if ((dir_x != 0 || dir_y != 0) && _is_valid_direction(x, y, dir_x, dir_y, player)) {
                return true;
            }
        }
    }
    return false;
}

bool game_engine::no_available_moves() {
    return (get_possible_moves_count(BLACK_PLAYER) == 0) && (get_possible_moves_count(WHITE_PLAYER) == 0);
}

int game_engine::count_players_scores(b_size &bp_scores, b_size &wp_scores) {
    size_t x;
    size_t y;
    char c;
    bp_scores = 0;
    wp_scores = 0;

    for (y = 0; y < this->game_board->get_rows(); ++y) {
        for (x = 0; x < this->game_board->get_cols(); ++x) {
            c = this->game_board->get_at(x, y);
            if (c == BLACK_PLAYER) {
                ++bp_scores;
            } else if (c == WHITE_PLAYER) {
                ++wp_scores;
            }
        }
    }
    return 0;
}

size_t game_engine::_make_move(size_t x, size_t y, char player) {
    int dir_x;
    int dir_y;
    size_t i;
    size_t steps;
    size_t count = 0;
    this->game_board->set_at(x, y, player);
    for (dir_x = -1; dir_x <= 1; ++dir_x) {
        for (dir_y = -1; dir_y <= 1; ++dir_y) {
            steps = _is_valid_direction(x, y, dir_x, dir_y, player);
            count += steps;
            for (i = 1; i <= steps; ++i) {
                this->game_board->set_at(x + i * dir_x, y + i * dir_y, player);
            }
        }
    }
    return count;
}

bool game_engine::process_move(b_size x, b_size y, player_code player) {
    if (!_is_valid_move(x, y, player)) {
        return false;
    }
    _make_move( x, y, player);
    return true;
}

File: sp/reversi/client/composeApp/src/commonMain/kotlin/com/vpavlov/ups/reversi/client/game/GameEngine.kt
package com.vpavlov.ups.reversi.client.game

import com.vpavlov.ups.reversi.client.domains.game.MoveCoordinates
import com.vpavlov.ups.reversi.client.domains.game.PlayerCode

class GameEngine(
    gameConfig: GameConfig,
    boardCells: List<Int>? = null
) {

    val board: Board = Board(rows = gameConfig.boardWidth, cols = gameConfig.boardHeight)

    init{
        //initializedBoard(DEFAULT_INIT_X, DEFAULT_INIT_Y, boardCells)
        if (boardCells != null){
            board.copyValues(boardCells)
        }

    }

    private fun initializedBoard(initX: Int, initY: Int, boardCells: List<Int>?){
        if (boardCells == null){
            board.setAt(initX, initY, PlayerCode.WHITE_PLAYER)
            board.setAt(initX+ 1, initY, PlayerCode.BLACK_PLAYER)
            board.setAt(initX, initY + 1, PlayerCode.BLACK_PLAYER)
            board.setAt(initX + 1, initY + 1, PlayerCode.WHITE_PLAYER)
        } else {
            board.copyValues(boardCells)
        }
    }

    fun getPossibleMovesCount(player: PlayerCode): Int {
        val (movesCount, _) = getPossibleMoves(player)
        return movesCount
    }

    fun getPossibleMoves(player: PlayerCode): Pair<Int, Array<Boolean>> {
        val moves = Array(board.rows * board.cols) { false }
        var movesCount = 0;
        for (x in 0..<board.cols) {
            for (y in 0..<board.rows) {
                if (isValidMove(MoveCoordinates(x, y), player)) {
                    moves[y * board.cols + x] = true
                    ++movesCount
                }
            }
        }
        return Pair(movesCount, moves)
    }

    fun isValidMove(move: MoveCoordinates, player: PlayerCode): Boolean {
        if (move.x !in 0..<board.cols || move.y !in 0..<board.rows) {
            return false
        }
        if (board.getAt(move.x, move.y) != PlayerCode.NO_PLAYER || player == PlayerCode.NO_PLAYER) {
            return false
        }
        /*Check throw all possible directions*/
        for (dirX in -1..1) {
            for (dirY in -1..1) {
                if (countSteps(move.x, move.y, dirX, dirY, player) > 0) {
                    return true
                }
            }
        }
        return false
    }

    private fun countSteps(x: Int, y: Int, dirX: Int, dirY: Int, player: PlayerCode): Int {
        val opponent = player.getOpponent()
        var steps = 1
        if (dirX == 0 && dirY == 0) {
            return 0
        }
        /*Check if move is in bounds and an opponent cell*/
        if ((x + dirX) !in 0..<board.cols || (y + dirY) !in 0..<board.rows || board.getAt(
                x + dirX,
                y + dirY
            ) != opponent
        ) {
            return 0
        }
        var posX = x + steps * dirX
        var posY = y + steps * dirY
        /*Check specified direction for a valid sequence of opponent cells*/
        while (posX in 0..<board.cols && posY in 0..<board.rows) {
            val cell = board.getAt(posX, posY)
            if (cell == PlayerCode.NO_PLAYER) {
                return 0
            }
            if (cell == player) {
                return steps
            }
            ++steps
            posX = x + steps * dirX
            posY = y + steps * dirY
        }
        return 0
    }

    fun countPlayersScores(): Pair<Int, Int> {
        var bpScores = 0
        var wpScores = 0
        for (x in 0..<board.cols) {
            for (y in 0..<board.rows) {
                val cell = board.getAt(x, y)
                when (cell) {
                    PlayerCode.BLACK_PLAYER -> ++bpScores
                    PlayerCode.WHITE_PLAYER -> ++wpScores
                    PlayerCode.NO_PLAYER -> {}
                }
            }
        }

        return Pair(wpScores, bpScores)
    }

    private fun makeMove(x: Int, y: Int, player: PlayerCode): Int {
        var steps = 0
        for (dirX in -1..1) {
            for (dirY in -1..1) {
                val tmpSteps = countSteps(x, y, dirX, dirY, player)
                steps += tmpSteps
                for (i in 1..tmpSteps) {
                    board.setAt(x + i * dirX, y + i * dirY, player)
                }
            }
        }
        return steps
    }

    fun processMove(move: MoveCoordinates, player: PlayerCode): Boolean {
        if (!isValidMove(move, player)) {
            return false
        }
        makeMove(move.x, move.y, player)
        return true
    }


}